spek-ide-plugin/intellij-common/src/main/kotlin/org/spekframework/intellij/domain/descriptors.kt

```diff
@@ -0,0 +1,226 @@
+}
+
+private val SPEK_CLASSES = listOf(
+    "org.spekframework.spek2.Spek"
+)
+
+private val SYNONYM_CLASSES = listOf(
+    "org.spekframework.spek2.meta.Synonym"
+)
+
+private val DESCRIPTIONS_CLASSES = listOf(
+    "org.spekframework.spek2.meta.Descriptions"
+)
+
+object ScopeDescriptorCache {
+    private val cache = ConcurrentHashMap<String, Pair<Long, ScopeDescriptor.Group>>()
+    private val index = ConcurrentHashMap<String, ScopeDescriptor>()
+
+    fun findDescriptor(path: Path): ScopeDescriptor? {
+        return index.getOrDefault(path.serialize(), null)
+    }
+
+    fun toDescriptor(callExpression: KtCallExpression): ScopeDescriptor? {
+        val context = fetchSynonymContext(callExpression)
+        if (context != null) {
+            return PsiTreeUtil.getParentOfType(callExpression, KtClassOrObject::class.java)?.let {
+                toDescriptor(it)?.findDescriptorForElement(callExpression)
+            }
+        }
+        return null
+    }
+
+    fun toDescriptor(clz: KtClassOrObject): ScopeDescriptor.Group? {
+        if (!isSpekSubclass(clz) || clz.isAbstract()) {
+            return null
+        }
+        val fqName = checkNotNull(clz.fqName).asString()
+
+        var cached = cache.getOrDefault(fqName, null)
+
+        if (cached == null || cached.first < clz.modificationStamp) {
+            cached = clz.modificationStamp to buildDescriptor(clz)
+            cache[fqName] = cached
+        }
+
+        return checkNotNull(cached).second
+    }
+
+    private fun buildDescriptor(clz: KtClassOrObject): ScopeDescriptor.Group {
+        val info = KtClassInfoUtil.createClassLikeInfo(clz)
+        val builder = PathBuilder()
+        if (info.containingPackageFqName.asString().isNotBlank()) {
+            builder.append(info.containingPackageFqName.asString())
+        }
+        val path = builder
+            .append(checkNotNull(clz.fqName).shortName().asString())
+            .build()
+
+        val superTypeCall = clz.superTypeListEntries.filterIsInstance<KtSuperTypeCallEntry>()
+            .firstOrNull()
+
+        val children = mutableListOf<ScopeDescriptor>()
+        if (superTypeCall != null) {
+            val lambda = superTypeCall.valueArguments.map { it.getArgumentExpression() }
+                .filterIsInstance<KtLambdaExpression>()
+                .firstOrNull()
+
+            if (lambda != null) {
+                lambda.bodyExpression?.let { body ->
+                    children.addAll(buildScopes(path, body))
+                }
+            }
+        }
+        return ScopeDescriptor.Group(path, clz, children.toList()).apply {
+            index[path.serialize()] = this
+        }
+    }
+
+    private fun buildScopes(parent: Path, block: KtBlockExpression): List<ScopeDescriptor> {
+        val callExpressions = PsiTreeUtil.getChildrenOfTypeAsList(block, KtCallExpression::class.java)
+        val scopes = mutableListOf<ScopeDescriptor>()
+
+        callExpressions.forEach { callExpression ->
```

Comment:
I'd turn this into a stream and then use `filter()` `mapNotNull` `flatMap` etc as I find it very hard to read this way since it's a mixture of both